File: Chap-4/multi-class/one_versus_all_demo.py

import numpy as np
import matplotlib.pyplot as plt


# load the data
def load_data():
    # load data
    data = np.matrix(np.genfromtxt('4class_data.csv', delimiter=','))
    x = np.asarray(data[:,0:2])
    temp = np.shape(x)
    temp = np.ones((temp[0],1))
    X = np.concatenate((temp,x),1)
    X = X.T
    y = np.asarray(data[:,2])
    y.shape = (np.size(y),1)

    # return needed variables
    return X,y

###### ML Algorithm functions ######
# learn all C separators
def learn_separators(X,y):
    W = []
    num_classes = np.size(np.unique(y))
    for i in range(0,num_classes):
        # prepare temporary C vs notC probem labels
        y_temp = np.copy(y)
        ind = np.argwhere(y_temp == (i+1))
        ind = ind[:,0]
        ind2 = np.argwhere(y_temp != (i+1))
        ind2 = ind2[:,0]
        y_temp[ind] = 1
        y_temp[ind2] = -1
        # run descent algorithm to classify C vs notC problem
        w = newtons_method(np.random.randn(3,1),X,y_temp)
        W.append(w)
    W = np.asarray(W)
    W.shape = (num_classes,3)
    W = W.T
    return W

# run newton's method
def newtons_method(w0,X,y):
    ## uncomment to record objective value at each iteration to check that algorithm is working
    # obj_path = []
    # obj = calculate_obj(w0,X,y)
    # obj_path.append(obj)
    w = w0

    # start gradient descent loop
    H = np.dot(np.diag(y[:,0]),X.T)
    s = np.shape(y)
    s = s[0]
    l = np.ones((s,1))
    grad = 1
    iter = 1
    max_its = 100
    while np.linalg.norm(grad) > 10**(-5) and iter <= max_its:
        # compute gradient
        temp = 1/(1 + my_exp(np.dot(H,w)))
        grad = - np.dot(H.T,temp)

        # compute Hessian
        g = temp*(l - temp)
        hess = np.dot(np.dot(X,np.diag(g[:,0])),X.T)

        # take Newton step = solve Newton system
        temp = np.dot(hess,w) - grad
        w = np.dot(np.linalg.pinv(hess),temp)
        # w = linalg.solve(hess, dot(hess,w) - grad)

        # update path containers
        # obj = calculate_obj(w,X,y)
        # obj_path.append(obj)
        iter+= 1

    # show final average gradient norm for sanity check
    s = np.dot(grad.T,grad)/np.size(grad)
    s = 'The final average norm of the gradient = ' + str(float(s[0])) + ' in ' + str(iter) + ' iterations of newtons method'
    print(s)

    # # for use in testing if algorithm minimizing/converging properly
    # obj_path = asarray(obj_path)
    # obj_path.shape = (iter,1)
    # plot(asarray(obj_path))
    # show()

    return w

# calculate the objective value for a given input weight w
def calculate_obj(w,X,y):
    obj = np.log(1 + my_exp(-y*np.dot(X.T,w)))
    obj = obj.sum()
    return obj

# avoid overflow when using exp - just cutoff after arguments get too large/small
def my_exp(u):
    s = np.argwhere(u > 100)
    t = np.argwhere(u < -100)
    u[s] = 0
    u[t] = 0
    u = np.exp(u)
    u[t] = 1
    return u

# calculate the objective value for a given input weight w
def calculate_obj(w):
    obj = np.log(1 + my_exp(-y*np.dot(X.T,w)))
    obj = obj.sum()
    return obj

# avoid overflow when using exp - just cutoff after arguments get too large/small
def my_exp(u):
    s = np.argwhere(u > 100)
    t = np.argwhere(u < -100)
    u[s] = 0
    u[t] = 0
    u = np.exp(u)
    u[t] = 1
    return u

###### plotting functions #######

def plot_data_and_subproblem_separators(X,y,W):
    # initialize figure, plot data, and dress up panels with axes labels etc.,
    num_classes = np.size(np.unique(y))
    color_opts = np.array([[1,0,0.4], [ 0, 0.4, 1],[0, 1, 0.5],[1, 0.7, 0.5],[0.7, 0.6, 0.5]])
    f,axs = plt.subplots(1,num_classes + 1,facecolor = 'white')

    r = np.linspace(0,1,150)
    for a in range(0,num_classes):
        # color current class
        axs[a].scatter(X[1,],X[2,], s = 30,color = '0.75')
        s = np.argwhere(y == a+1)
        s = s[:,0]
        axs[a].scatter(X[1,s],X[2,s], s = 30,color = color_opts[a,:])
        axs[num_classes].scatter(X[1,s],X[2,s], s = 30,color = color_opts[a,:])

        # draw subproblem separator
        z = -W[0,a]/W[2,a] - W[1,a]/W[2,a]*r
        axs[a].plot(r,z,'-k',linewidth = 2,color = color_opts[a,:])

        # dress panel correctly
        axs[a].set_xlim(0,1)
        axs[a].set_ylim(0,1)
        axs[a].set(aspect = 'equal')
    axs[num_classes].set(aspect = 'equal')

    return axs

# fuse individual subproblem separators into one joint rule
def plot_joint_separator(W,axs,num_classes):
    r = np.linspace(0,1,300)
    s,t = np.meshgrid(r,r)
    s = np.reshape(s,(np.size(s),1))
    t = np.reshape(t,(np.size(t),1))
    h = np.concatenate((np.ones((np.size(s),1)),s,t),1)
    f = np.dot(W.T,h.T)
    z = np.argmax(f,0)
    f.shape = (np.size(f),1)
    s.shape = (np.size(r),np.size(r))
    t.shape = (np.size(r),np.size(r))
    z.shape = (np.size(r),np.size(r))

    for i in range(0,num_classes + 1):
        axs[num_classes].contour(s,t,z,(i + 0.5,i + 0.5),colors = 'k',linewidths = 2.25)

def main():
    # load the data
    X,y = load_data()

    # learn all C vs notC separators
    W = learn_separators(X,y)

    # plot data and each subproblem 2-class separator
    axs = plot_data_and_subproblem_separators(X,y,W)

    # plot fused separator
    plot_joint_separator(W,axs,np.size(np.unique(y)))

    plt.show()

main()

File: Chap-4/multi-class/one_versus_all_demo.m
function one_versus_all_demo()
clear all

% load data
red =  [ 1 0 .4];
blue =  [ 0 .4 1];
green = [0 1 0.5];
cyan = [1 0.7 0.5];
grey = [.7 .6 .5];
colors = [red;blue;green;cyan;grey];     % 4 maximum classes
[X,y] = load_data();

% find separators
W = learn_separators(X,y);

% plot separators one-at-a-time
figure(1)
plot_one_ata_time(W,X,y)

% % plot the max(..) separator
% figure(2)
% plot_max_separator(W,X,y)
    
%%%%%%%%%%%%%%%%% functions %%%%%%%%%%%%%%%%%%%

%%% learn separator for each class %%%
function W = learn_separators(X,y)  
    class_labels = unique(y);           % class labels
    num_classes = length(unique(y));
    W = [];         % container for all the weights to learn
    for i = 1:num_classes
        % setup temporary labels for one-vs-all classification       
        y_temp = y;
        class = class_labels(i);
        ind = find(y_temp == class);
        ind2 = find(y_temp ~= class);
        y_temp(ind) = 1;
        y_temp(ind2) = -1;
        
        % run newton's method and store resulting weights
        w = softmax_newtons_method(X,y_temp,20);
        W = [W, w];
    end

end

function w = softmax_newtons_method(X,y,max_its)
    % precomputations
    H = bsxfun(@times,X',y);
    w = randn(size(X,1),1)/size(X,1);
    
    %%% main %%%
    k = 1;
    while k < max_its
        
        % prep gradient for logistic objective
        r = sigmoid(-H*w);
        g = r.*(1 - r);
        grad = -H'*r;
        hess = bsxfun(@times,g,X')'*X';

        % take Newton step
        temp = hess*w - grad;
        w = pinv(hess)*temp;
        k = k + 1;
    end
    
    function t = sigmoid(z)
        t = 1./(1 + exp(-z));
    end 
    
end

%%% plot full dataset %%%
function plot_data(X,y)
    % how many classes in the data?  4 maximum for this toy.
    class_labels = unique(y);           % class labels
    num_classes = length(class_labels);

    % plot data
    for i = 1:num_classes
        class = class_labels(i);
        ind = find(y == class);
        hold on
        scatter(X(2,ind),X(3,ind),'Linewidth',2,'Markeredgecolor',colors(i,:),'markerFacecolor','none');

        hold on
        scatter(X(2,ind),X(3,ind),'Linewidth',2,'Markeredgecolor',colors(i,:),'markerFacecolor','none');
    end
    axis([0 1 0 1])
    axis square
    box on
end

%%% plot max separator %%%
function plot_max_separator(W,X,y,deg)
    % plot data
    subplot(1,3,1)
    plot_data(X,y)
    subplot(1,3,2)
    plot_data(X,y)
    subplot(1,3,3)
    plot_data(X,y)

    %%% plot all linear separators %%%
    subplot(1,3,2)
    num_classes = length(unique(y));
    x = [0:0.01:1];
    for j = 1:num_classes
        hold on
        w = W(:,j);    
        plot (x,(-w(1)-w(2)*x)/w(3),'Color',colors(j,:),'linewidth',2);
    end

    %%% generate max-separator surface %%%
    s = [0:0.0005:1];
    [s1,s2] = meshgrid(s,s);
    s1 = reshape(s1,numel(s1),1);
    s2 = reshape(s2,numel(s2),1);
    
    % compute criteria for each point in the range [0,1] x [0,1]
    p = [ones(size(s1(:),1),1),s1(:), s2(:)];
    f = W'*p';         
    [f,z] = max(f,[],1);
    
    % fill in appropriate regions with class colors
    co = zeros(size(p,1),3);
    ind = find(z == 1);
    co(ind,:) = repmat(red,length(ind),1);
    ind = find(z == 2);
    co(ind,:) = repmat(blue,length(ind),1);
    ind = find(z == 3);
    co(ind,:) = repmat(green,length(ind),1);
    ind = find(z == 4);   
    co(ind,:) = repmat(cyan,length(ind),1);  
    ind = find(z == 5);
    co(ind,:) = repmat(grey,length(ind),1);  
    hold on

    %%% plot colored region
     %scatter(s1,s2,1,co)
 
    
    % produce decision boundary
    s1 = reshape(s1,[length(s),length(s)]);
    s2 = reshape(s2,[length(s),length(s)]);   
    z = reshape(z,[length(s),length(s)]);   
    num_classes = length(unique(z));
    subplot(1,3,3)
    for i = 1:num_classes - 1
       hold on
       contour(s1,s2,z,[i + 0.5,i + 0.5],'Color','k','LineWidth',2)
    end
    
    % make plot real nice lookin'
    for i = 1:3
        subplot(1,3,i)
        axis([0 1 0 1])
        axis square
        xlabel('x_1','FontName','cmmi9','Fontsize',18)
        ylabel('x_2','FontName','cmmi9','Fontsize',18)
        set(get(gca,'YLabel'),'Rotation',0)
        set(gcf,'color','w');
    end
end

%%% plot learned separators one at a time
function plot_one_ata_time(W,X,y)

    %%% plot one separator at a time %%%
    class_labels = unique(y);           % class labels
    num_classes = length(unique(y));
    x = [0:0.01:1];

    for i = 1:num_classes
        subplot(1,num_classes+1,i)
        w = W(:,i);    
        plot (x,(-w(1)-w(2)*x)/w(3),'Color',colors(i,:),'linewidth',2);

        class = class_labels(i);
        ind = find(y == class);
        hold on
        scatter(X(2,ind),X(3,ind),'Linewidth',2,'Markeredgecolor',colors(i,:),'markerFacecolor','none');
        hold on
        ind = find(y ~= class);
        scatter(X(2,ind),X(3,ind),'Linewidth',2,'Markeredgecolor',[0.5 0.5 0.5],'markerFacecolor','none');
        axis([0 1 0 1])
        axis square
        box on

    end
    
    %%% plot fused separator %%%
    subplot(1,num_classes+1,num_classes+1)
    plot_data(X,y)
    
    % generate max-separator surface 
    s = [0:0.0005:1];
    [s1,s2] = meshgrid(s,s);
    s1 = reshape(s1,numel(s1),1);
    s2 = reshape(s2,numel(s2),1);
    
    % compute criteria for each point in the range [0,1] x [0,1]
    p = [ones(size(s1(:),1),1),s1(:), s2(:)];
    f = W'*p';         
    [f,z] = max(f,[],1);
    
    % fill in appropriate regions with class colors
    co = zeros(size(p,1),3);
    ind = find(z == 1);
    co(ind,:) = repmat(red,length(ind),1);
    ind = find(z == 2);
    co(ind,:) = repmat(blue,length(ind),1);
    ind = find(z == 3);
    co(ind,:) = repmat(green,length(ind),1);
    ind = find(z == 4);
    co(ind,:) = repmat(cyan,length(ind),1);  
    ind = find(z == 5);
    co(ind,:) = repmat(grey,length(ind),1);  
    hold on
    
    % produce decision boundary
    s1 = reshape(s1,[length(s),length(s)]);
    s2 = reshape(s2,[length(s),length(s)]);   
    z = reshape(z,[length(s),length(s)]);   
    num_classes = length(unique(z));
    for i = 1:num_classes - 1
       hold on
       contour(s1,s2,z,[i + 0.5,i + 0.5],'Color','k','LineWidth',2)
    end

    % make subplots real nice lookin'
    for i = 1:num_classes + 1
        subplot(1,num_classes + 1,i)
        axis([0 1 0 1])
        axis square
        xlabel('x_1','FontName','cmmi9','Fontsize',18)
        ylabel('x_2','FontName','cmmi9','Fontsize',18)
        set(get(gca,'YLabel'),'Rotation',0)
    end
    set(gcf,'color','w');
end

%%% load data %%%
function [X,y] = load_data()
    % load data from file
    data = csvread('4class_data.csv');
    X = data(:,1:end - 1);
    y = data(:,end);
    X = [ones(size(X,1),1), X];
    X = X';
end
end